### ISSUE TITLE
Add client support for server push

### ISSUE THREAD

**Issue by twdkeule** [lucas-clemente/quic-go#770]
I tried a setup where a Caddy proxy and a Caddy upstream server were using QUIC,
This fails as soon as the server pushes a resource to the client:
```
Sending PUSH_PROMISE for target '/dash.all.debug.js' on stream 3, promised on stream 2
2017/07/26 13:39:32.775762 -> Sending packet 0x9 (175 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.775782 	-> &frames.StreamFrame{StreamID: 3, FinBit: false, Offset: 0x0, Data length: 0x9e, Offset + Data length: 0x9e}
2017/07/26 13:39:32.779200 Responding with 200
2017/07/26 13:39:32.779292 -> Sending packet 0xa (141 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779313 	-> &frames.StreamFrame{StreamID: 3, FinBit: false, Offset: 0x9e, Data length: 0x7a, Offset + Data length: 0x118}
2017/07/26 13:39:32.779454 -> Sending packet 0xb (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779475 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x0, Data length: 0x535, Offset + Data length: 0x535}
2017/07/26 13:39:32.779513 -> Sending packet 0xc (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779537 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x535, Data length: 0x533, Offset + Data length: 0xa68}
2017/07/26 13:39:32.779568 -> Sending packet 0xd (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779611 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0xa68, Data length: 0x533, Offset + Data length: 0xf9b}
2017/07/26 13:39:32.779642 -> Sending packet 0xe (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779660 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0xf9b, Data length: 0x533, Offset + Data length: 0x14ce}
2017/07/26 13:39:32.779690 -> Sending packet 0xf (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779708 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x14ce, Data length: 0x533, Offset + Data length: 0x1a01}
2017/07/26 13:39:32.779739 -> Sending packet 0x10 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779756 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x1a01, Data length: 0x533, Offset + Data length: 0x1f34}
2017/07/26 13:39:32.779785 -> Sending packet 0x11 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779802 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x1f34, Data length: 0x533, Offset + Data length: 0x2467}
2017/07/26 13:39:32.779831 -> Sending packet 0x12 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779848 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x2467, Data length: 0x533, Offset + Data length: 0x299a}
2017/07/26 13:39:32.779882 -> Sending packet 0x13 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779899 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x299a, Data length: 0x533, Offset + Data length: 0x2ecd}
2017/07/26 13:39:32.779927 -> Sending packet 0x14 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.779944 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x2ecd, Data length: 0x533, Offset + Data length: 0x3400}
2017/07/26 13:39:32.780094 -> Sending packet 0x15 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780116 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x3400, Data length: 0x533, Offset + Data length: 0x3933}
2017/07/26 13:39:32.780158 -> Sending packet 0x16 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780176 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x3933, Data length: 0x533, Offset + Data length: 0x3e66}
2017/07/26 13:39:32.780207 -> Sending packet 0x17 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780225 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x3e66, Data length: 0x533, Offset + Data length: 0x4399}
2017/07/26 13:39:32.780264 -> Sending packet 0x18 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780282 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x4399, Data length: 0x533, Offset + Data length: 0x48cc}
2017/07/26 13:39:32.780317 -> Sending packet 0x19 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780334 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x48cc, Data length: 0x533, Offset + Data length: 0x4dff}
2017/07/26 13:39:32.780367 -> Sending packet 0x1a (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780392 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x4dff, Data length: 0x533, Offset + Data length: 0x5332}
2017/07/26 13:39:32.780433 -> Sending packet 0x1b (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780450 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x5332, Data length: 0x533, Offset + Data length: 0x5865}
2017/07/26 13:39:32.780487 -> Sending packet 0x1c (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780504 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x5865, Data length: 0x533, Offset + Data length: 0x5d98}
2017/07/26 13:39:32.780533 -> Sending packet 0x1d (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780636 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x5d98, Data length: 0x533, Offset + Data length: 0x62cb}
2017/07/26 13:39:32.780672 -> Sending packet 0x1e (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780698 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x62cb, Data length: 0x533, Offset + Data length: 0x67fe}
2017/07/26 13:39:32.780729 -> Sending packet 0x1f (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780746 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x67fe, Data length: 0x533, Offset + Data length: 0x6d31}
2017/07/26 13:39:32.780786 -> Sending packet 0x20 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.780804 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x6d31, Data length: 0x533, Offset + Data length: 0x7264}
2017/07/26 13:39:32.780836 -> Sending packet 0x21 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.781029 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x7264, Data length: 0x533, Offset + Data length: 0x7797}
2017/07/26 13:39:32.781088 -> Sending packet 0x22 (1350 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.781106 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x7797, Data length: 0x533, Offset + Data length: 0x7cca}
2017/07/26 13:39:32.781181 -> Sending packet 0x23 (841 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.781200 	-> &frames.StreamFrame{StreamID: 2, FinBit: false, Offset: 0x7cca, Data length: 0x336, Offset + Data length: 0x8000}
2017/07/26 13:39:32.781289 -> Sending packet 0x24 (20 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.785198 	-> &frames.BlockedFrame{StreamID:0x2}
2017/07/26 13:39:32.876116 <- Reading packet 0x9 (32 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.876170 	<- &frames.StopWaitingFrame{LeastUnacked: 0x9}
2017/07/26 13:39:32.876201 	<- &frames.AckFrame{LargestAcked: 0x9, LowestAcked: 0x8, AckRanges: []frames.AckRange(nil), DelayTime: 54Âµs}
2017/07/26 13:39:32.876261 <- Reading packet 0xa (49 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.876295 	<- &frames.ConnectionCloseFrame{ErrorCode:0x38, ReasonPhrase:"not a headers frame"}
2017/07/26 13:39:32.876324 -> Sending packet 0x25 (24 bytes) for connection 4c0eabfc5dac0c83, forward-secure
2017/07/26 13:39:32.876340 	-> &frames.StopWaitingFrame{LeastUnacked: 0xa, PacketNumberLen: 0x2}
2017/07/26 13:39:32.876356 	-> &frames.AckFrame{LargestAcked: 0xa, LowestAcked: 0x9, AckRanges: []frames.AckRange(nil), DelayTime: 39.277Âµs}
2017/07/26 13:39:32.876439 Closing session with error: InvalidHeadersStreamData: not a headers frame
2017/07/26 13:39:32.876562 TDK: Pushed /dash.all.debug.js
2017/07/26 13:39:32.876578 TDK:     Resource: /thesis.js
2017/07/26 13:39:32.876630 TDK: Pushed /thesis.js
2017/07/26 13:39:32.876711 Responding with 200
2017/07/26 13:39:32.876738 could not write h2 header: InvalidHeadersStreamData: not a headers frame
```

**Comment by twdkeule** [comment]
As @marten-seemann said, this should best follow the standard library (https://github.com/golang/go/issues/18594).

**Comment by marten-seemann** [comment]
Closed for the same reason as #372.
